src-input/duk_bi_encoding.c

```diff
@@ -0,0 +1,348 @@
+	duk__encode_context *enc_ctx;
+
+	enc_ctx = (duk__encode_context *) udata;
+	if (codepoint > 0x10ffffL) {
+		/* not a valid UTF-8 codepoint */
+		enc_ctx->out += duk_unicode_encode_xutf8(DUK_UNICODE_CP_REPLACEMENT_CHARACTER, enc_ctx->out);
+	} else if (codepoint >= 0xd800L && codepoint <= 0xdbffL) {
+		/* high surrogate */
+		if (enc_ctx->lead != 0x0000L) {
+			/* consecutive high surrogates, consider first one unpaired */
+			enc_ctx->out += duk_unicode_encode_xutf8(DUK_UNICODE_CP_REPLACEMENT_CHARACTER, enc_ctx->out);
+		}
+		enc_ctx->lead = codepoint;
+	} else if (codepoint >= 0xdc00L && codepoint <= 0xdfffL) {
+		/* low surrogate */
+		if (enc_ctx->lead != 0x0000L) {
+			codepoint = 0x010000L | ((enc_ctx->lead - 0xd800L) << 10) | (codepoint - 0xdc00L);
+			enc_ctx->out += duk_unicode_encode_xutf8(codepoint, enc_ctx->out);
+			enc_ctx->lead = 0x0000L;
+		} else {
+			/* unpaired low surrogate */
+			enc_ctx->out += duk_unicode_encode_xutf8(DUK_UNICODE_CP_REPLACEMENT_CHARACTER, enc_ctx->out);
+		}
+	} else {
+		if (enc_ctx->lead != 0x0000L) {
+			/* unpaired high surrogate */
+			enc_ctx->out += duk_unicode_encode_xutf8(DUK_UNICODE_CP_REPLACEMENT_CHARACTER, enc_ctx->out);
+			enc_ctx->lead = 0x0000L;
+		}
+		enc_ctx->out += duk_unicode_encode_xutf8(codepoint, enc_ctx->out);
+	}
+}
+
+DUK_INTERNAL duk_ret_t duk_bi_textencoder_constructor(duk_context *ctx) {
+	/*
+	 *  TextEncoder currently requires no persistent state, so this constructor
+	 *  does nothing on purpose.
+	 */
+
+	if (!duk_is_constructor_call(ctx)) {
+        duk_error(ctx, DUK_ERR_TYPE_ERROR, "constructor must use 'new'");
+    }
+	return 0;
+}
+
+DUK_INTERNAL duk_ret_t duk_bi_textencoder_prototype_encoding_getter(duk_context *ctx) {
+	duk_push_string(ctx, "utf-8");
+	return 1;
+}
+
+DUK_INTERNAL duk_ret_t duk_bi_textencoder_prototype_encode(duk_context *ctx) {
+	duk__encode_context enc_ctx;
+	duk_size_t len;
+	duk_uint8_t *output;
+
+	if (duk_get_top(ctx) >= 1) {
+		duk_dup(ctx, 0);
+		duk_to_string(ctx, -1);
+	} else {
+		duk_push_string(ctx, "");
+	}
+
+	len = duk_get_length(ctx, -1);
+	if (len >= DUK_HBUFFER_MAX_BYTELEN / 3) {
+		duk_error(ctx, DUK_ERR_TYPE_ERROR, "input string too long");
+	}
+
+	/* XXX: The buffer allocation strategy here (3x input length) is quite
+	 * inefficient.  Maybe switch to a chunk-based strategy?
+	 */
+
+	output = (duk_uint8_t *) duk_push_dynamic_buffer(ctx, len * 3);
+	enc_ctx.lead = 0x0000L;
+	enc_ctx.out = output;
+	duk_decode_string(ctx, -2, duk__utf8_encode_char, (void *) &enc_ctx);
+	if (enc_ctx.lead != 0x0000L) {
+		/* unpaired high surrogate at end of string */
+		enc_ctx.out += duk_unicode_encode_xutf8(DUK_UNICODE_CP_REPLACEMENT_CHARACTER, enc_ctx.out);
+	}
+#if defined(DUK_USE_BUFFEROBJECT_SUPPORT)
+	duk_push_buffer_object(ctx, -1, 0, (duk_size_t) (enc_ctx.out - output), DUK_BUFOBJ_UINT8ARRAY);
+#else
+	/* shrink buffer to include only the encoded text */
+	duk_resize_buffer(ctx, -1, (duk_size_t) (enc_ctx.out - output));
```

Comment:
Right, the practical alternative would be to push a second fixed buffer for the final result. Not very nice either which is why a dynamic buffer might be the best solution here.

In any case the resulting buffer object will be different than a buffer object created using e.g. `new ArrayBuffer()` or `new Uint8Array()` which are always backed by fixed buffers now. Just thought this is noteworthy if you weren't aware of it, as it's a change to the current behavior for Ecmascript-binding-created buffer objects.

There may be some documentation that needs to be updated to reflect this; it shouldn't have a large impact in any case because C code can always create buffers with fixed, dynamic, or external backing.

Re: memory safety: Duktape won't do that, but it's critical to prove there can be no access from finalizers etc. I think that should be a safe assumption here but it's easy to violate in the general case.